x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/action/TransportPutLifecycleAction.java

```diff
@@ -77,55 +80,76 @@ protected void masterOperation(Task task, Request request, ClusterState state, A
         // same context, and therefore does not have access to the appropriate security headers.
         Map<String, String> filteredHeaders = ClientHelper.filterSecurityHeaders(threadPool.getThreadContext().getHeaders());
         LifecyclePolicy.validatePolicyName(request.getPolicy().getName());
-        List<Phase> phasesDefiningSearchableSnapshot = request.getPolicy().getPhases().values().stream()
-            .filter(phase -> phase.getActions().containsKey(SearchableSnapshotAction.NAME))
+        List<SearchableSnapshotAction> searchableSnapshotActions = request.getPolicy().getPhases().values().stream()
+            .map(phase -> (SearchableSnapshotAction) phase.getActions().get(SearchableSnapshotAction.NAME))
+            .filter(Objects::nonNull)
             .collect(Collectors.toList());
-        if (phasesDefiningSearchableSnapshot.isEmpty() == false) {
-            if (SEARCHABLE_SNAPSHOT_FEATURE.checkWithoutTracking(licenseState) == false) {
-                throw new IllegalArgumentException("policy [" + request.getPolicy().getName() + "] defines the [" +
-                    SearchableSnapshotAction.NAME + "] action but the current license is non-compliant for [searchable-snapshots]");
+        // check license level for searchable snapshots
+        if (searchableSnapshotActions.isEmpty() == false && SEARCHABLE_SNAPSHOT_FEATURE.checkWithoutTracking(licenseState) == false) {
+            throw new IllegalArgumentException("policy [" + request.getPolicy().getName() + "] defines the [" +
+                SearchableSnapshotAction.NAME + "] action but the current license is non-compliant for [searchable-snapshots]");
+        }
+        // make sure any referenced snapshot repositories exist
+        for (SearchableSnapshotAction action : searchableSnapshotActions) {
+            String repository = action.getSnapshotRepository();
+            if (state.metadata().custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY).repository(repository) == null) {
+                throw new IllegalArgumentException("no such repository [" + repository + "]");
+            }
+        }
+
+        List<WaitForSnapshotAction> waitForSnapshotActions = request.getPolicy().getPhases().values().stream()
+            .map(phase -> (WaitForSnapshotAction) phase.getActions().get(WaitForSnapshotAction.NAME))
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        // make sure any referenced snapshot lifecycle policies exist
+        for (WaitForSnapshotAction action : waitForSnapshotActions) {
+            String policy = action.getPolicy();
+            if (state.metadata().custom(SnapshotLifecycleMetadata.TYPE, SnapshotLifecycleMetadata.EMPTY)
+                .getSnapshotConfigurations().get(policy) == null) {
+                throw new IllegalArgumentException("no such snapshot lifecycle policy [" + policy + "]");
```

Comment:
Should this validation happen in the update task? 

I think so, because we're reading from the cluster state outside of master update thread, 
which could mean that when validation succeeds now, but when the update task is performed
it fails. 

Alternatively we can do this check both before submitting the update task and as part
of the update task. This is a cheap check and the likeliness of a repository being removed during 
time between the pre submit task validation logic and when the update task is being executed is small.

Reply:
Ahhhhhhhhhhhhhhh, hmmmm... I'll take a look at that. Good suggestion!

Reply:
Done, 05b34b8bbb7a87c2797c16e12840d5f0083fdadd